lib/compat/wordpress-6.0/block-editor-settings.php

```diff
@@ -34,49 +68,39 @@ function_exists( 'gutenberg_is_edit_site_page' ) &&
 	}
 
 	if ( 'other' === $context ) {
+		global $wp_version;
+		$is_wp_5_8 = version_compare( $wp_version, '5.8', '>=' ) && version_compare( $wp_version, '5.9', '<' );
+		$is_wp_5_9 = version_compare( $wp_version, '5.9', '>=' ) && version_compare( $wp_version, '6.0', '<' );
+
 		// Make sure the styles array exists.
 		// In some contexts, like the navigation editor, it doesn't.
 		if ( ! isset( $settings['styles'] ) ) {
 			$settings['styles'] = array();
 		}
 
+		// Remove existing global styles provided by core.
 		$styles_without_existing_global_styles = array();
 		foreach ( $settings['styles'] as $style ) {
 			if (
-				! isset( $style['__unstableType'] ) ||
-				// '__unstableType' is 'globalStyles' for WordPress 5.8 and 'presets' for WordPress 5.9.
-				//
-				// Note that styles classified as'theme', can be from the theme stylesheet
-				// or from the theme.json (the styles section).
-				// We are unable to identify which is which, so we can't remove and recreate those.
-				// Instead, we reload the theme.json styles from the plugin.
-				// Because they'll use the same selectors and load later,
-				// they'll have higher priority than core's.
-				//
-				// Theoretically, this approach with 'theme' styles could be problematic:
-				// if we remove style properties in the plugin, if selectors change, etc.
-				// We need to address this issue directly in core by alowing to identify
-				// styles coming from theme.json.
-				//
-				// A final note about 'theme' styles: this flag is used to identify theme
-				// styles that may need to be removed if the user toggles
-				// "Preferences > Use theme styles" in the preferences modal.
-				//
-				! in_array( $style['__unstableType'], array( 'globalStyles', 'presets' ), true )
+				( $is_wp_5_8 && ! gutenberg_is_global_styles_in_5_8( $style ) ) ||
+				( $is_wp_5_9 && ! gutenberg_is_global_styles_in_5_9( $style ) )
 			) {
 				$styles_without_existing_global_styles[] = $style;
 			}
 		}
 
+		// Recreate global styles.
 		$new_global_styles = array();
 		$presets           = array(
 			array(
 				'css'            => 'variables',
 				'__unstableType' => 'presets',
+				'isGlobalStyles' => true,
```

Comment:
`isGlobalStyles` is an existing flag used by the client to determine whether it needs to discard this and recreate them from scratch. We can change this to `__unstableSource` as we [talked](https://github.com/WordPress/wordpress-develop/pull/2542#discussion_r846092243) (also in the client) but given that one key already exists I thought we could as well just leverage it.

The only change in https://github.com/WordPress/wordpress-develop/pull/2542 could be set that key for the styles coming from global styles.

Reply:
cc @youknowriad thoughts?

Reply:
This actually fixes the bug for 5.9 (and still works for 5.8). If direction is good I'll add another check for 6.0 here https://github.com/WordPress/gutenberg/pull/40185/files#r848391697 and will also push this to the wordpress-develop PR https://github.com/WordPress/wordpress-develop/pull/2542

Reply:
We can use `__unstableSource` if we want, but then in the client we need to transform it to `isGlobalStyles: true`. Given there's already a flag that captures what we want to do, I thought to expand its use instead of creating a new (unstable) one. I'm also fine with using `__unstableSource`.